controller/konnect/reconciler_kongplugin.go

```diff
@@ -242,31 +251,187 @@ func (r *KongPluginReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+	return func(ownerRef metav1.OwnerReference) bool {
+		return ownerRef.Kind == "KongPlugin" &&
+			ownerRef.Name == kongPlugin.Name &&
+			ownerRef.UID == kongPlugin.UID
 	}
+}
+
+func deleteUnusedKongPluginBindings(
+	ctx context.Context,
+	logger logr.Logger,
+	clientWithNamespace client.Client,
+	kongPlugin *configurationv1.KongPlugin,
+	groupedCombinations map[types.NamespacedName][]Rel,
+	kongPluginBindings []configurationv1alpha1.KongPluginBinding,
+) error {
+	pluginBindingsToDelete := make(map[types.NamespacedName]configurationv1alpha1.KongPluginBinding)
+	for _, pb := range kongPluginBindings {
+		// If the KongPluginBinding has a deletion timestamp, do not delete it.
+		if !pb.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		// If the KongPluginBinding is unmanaged (created not using an annotation), skip it, do not delete it.
+		if !lo.ContainsBy(pb.OwnerReferences, ownerRefIsPlugin(kongPlugin)) {
+			continue
+		}
 
-	for _, b := range bindings {
-		serviceRef := b.Spec.Targets.ServiceReference
-		if serviceRef != nil &&
-			serviceRef.Group == configurationv1alpha1.GroupVersion.Group &&
-			serviceRef.Kind == "KongService" {
-			ret.byServiceName[serviceRef.Name] = append(ret.byServiceName[serviceRef.Name], b)
+		cpRef := pb.Spec.ControlPlaneRef
+		if cpRef == nil ||
+			cpRef.KonnectNamespacedRef == nil ||
+			cpRef.Type != configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef {
+			continue
 		}
 
-		routeRef := b.Spec.Targets.RouteReference
-		if routeRef != nil &&
-			routeRef.Group == configurationv1alpha1.GroupVersion.Group &&
-			routeRef.Kind == "KongRoute" {
-			ret.byRouteName[routeRef.Name] = append(ret.byRouteName[routeRef.Name], b)
+		combinations, ok := groupedCombinations[types.NamespacedName{
+			// TODO: implement cross namespace references
+			Namespace: pb.Namespace,
+			Name:      cpRef.KonnectNamespacedRef.Name,
+		}]
+		if !ok {
+			pluginBindingsToDelete[client.ObjectKeyFromObject(&pb)] = pb
+			continue
 		}
+
+		// If the konghq.com/plugins annotation is not present, it doesn't contain
+		// the plugin in question or the object referring to the plugin has a non zero deletion timestamp,
+		// we need to delete all the managed KongPluginBindings that reference the object.
+
+		serviceRef := pb.Spec.Targets.ServiceReference
+		routeRef := pb.Spec.Targets.RouteReference
+		switch {
+
+		case serviceRef != nil && routeRef != nil:
+			combinationFound := false
+			for _, rel := range combinations {
+				if rel.Service != serviceRef.Name &&
+					rel.Route != routeRef.Name {
+					continue
+				}
+				combinationFound = true
+				break
+			}
+
+			if !combinationFound {
+				pluginBindingsToDelete[client.ObjectKeyFromObject(&pb)] = pb
+				continue
+			}
+
+			s, serviceExists := getIfRefNotNil[*configurationv1alpha1.KongService](ctx, clientWithNamespace, serviceRef)
+			r, routeExists := getIfRefNotNil[*configurationv1alpha1.KongRoute](ctx, clientWithNamespace, routeRef)
```

Comment:
This could be moved up, out of the switch, same for other `getIfRefNotNil` calls.

Reply:
But then we could end up getting objects that we don't need, right? Do you think this (moving this up) is worth the potential unnecessary object lookup?

Reply:
Yeah, makes sense. Let's leave it.